### PR TITLE
Data point symbols

### PR DESCRIPTION
**Pull request by nightlyop**
Currently, all data points in line charts are displayed as circles. I need the possibility to use other symbols (eg. squares, stars, diamonds, "v", ...) to display the data points. Moreover I need to be able to specify the symbol for every data point. The reason therefore is, that my chart has to use the same system as an older one we're selling. In our case the kind of symbol indicates special observations made when the displayed value was measured.

I tried to make the Voronoi Container more generic and so far the symbols `circle` (default; same as before) and `rect` can be used.

`VoronoiContainer.jsx` replaces `VoronoiCircleContainer.jsx` because most of the symbols I'd like to create will share most of the code in here. Because of that I also renamed variables like `circleFill` to `symbolFill`.

What do you think about this approach?

### CODE REVIEW

**Comment by yang-wei** [comment]
@nightlyop Hi Chris thanks for the pr. Briefly looked at them and I like what you are adding to this library.

However first of all I don't really prefer `symbol` as the name. For me `marker` sounds like a better name. What do you think ?

**Comment by nightlyop** [comment]
@yang-wei You're right, `marker` is better. I'll change that.

**Comment by yang-wei** [comment]
> VoronoiContainer.jsx replaces VoronoiCircleContainer.jsx
agree.

But I saw code replication between VoronoiCircle and VoronoiRect. VoronoiContainer should not worry about what to rendering. It's job is to handle state(in our case).

So I will prefer the `switch` statement to go into VoronoiChild( currently VoronoiCircle ) and the voronoi child will decide what to render. Does this make sense?

**Comment by nightlyop** [comment]
Do you mean a component VoronoiContainer and a component VoronoiMarker and VoronoiMarker contains a `switch` to decide what to render? That would sound good for me as well. I'm a beginner and don't know what's better but I'm glad if I can improve it with your feedback ;-)

**Comment by MatthewHerbst** [comment]
I like this!

@yang-wei I actually think the switch statement is in the correct place. A VoronoiChild such as `VoronoiCircle` should have no knowledge of the other children such as `VoronoiRect`.

**Comment by yang-wei** [comment]
@nightlyop Yes VoronoiMarker will be a better name. 

What I see about VoronoiRect is it has a lot of code duplication with VoronoiCircle. That's what I want to avoid so I think the VoronoiMarker should contains the switch statement.

**Comment by nightlyop** [comment]
Ok, I'll rename everything and replace `symbol` by `marker`
After reading 
* [Smart and Dumb Components](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0)
* [Container Components](https://medium.com/@learnreact/container-components-c0e67432e005)

I also thinkt the switch statement is at the correct place. The components `VoronoiRect` and `VoronoiCircle` are indeed very similar. But I expect other kind of markers, for which there is no ready to use SVG component (eg. a star or a "v"), to be more different.

**Comment by yang-wei** [comment]
IMHO what make VoronoiRect and VoronoiCircle similar is the `<path>` component.
```
      <g>
        <path
          onMouseOver={this.props.handleMouseOver}
          onMouseLeave={this.props.handleMouseLeave}
          fill='white'
          d={this.props.voronoiPath} 
        />
       ...
      </g>
```
Because they are voronoi-ed so we need this component on every child of container component. Imagine that if we create another marker then we have one more `VoronoiNewMarker` file. 
Then one if we have to change our `<path>` api, we need to touch a lot of files. This is what I want to avoid. Let me know what do you think :smile:

**Comment by nightlyop** [comment]
How about putting that part in a mixin?

**Comment by yang-wei** [comment]
I don't see mixin is an alternative but if you don't mind to explain, can you elaborate?

**Comment by nightlyop** [comment]
A collegue told me that mixin is not a good solution because they won't be supported in the future. How about making a higher order component wrapping `path` (according to [Mixins Are Dead. Long Live Composition](https://medium.com/@dan_abramov/mixins-are-dead-long-live-higher-order-components-94a0d2f9e750))

**Comment by yang-wei** [comment]
No one knows the future of mixin but if you look at [react-future](https://github.com/reactjs/react-future/blob/master/01%20-%20Core%2F02%20-%20Mixins.js) or [decorator](http://raganwald.com/2015/06/26/decorators-in-es7.html) it's still possible to use mixin.

HIgher order component is the thing I've been thinking all the time but I still can come out with a good pattern.

**Comment by nightlyop** [comment]
To summarize the update:
- I renamed everything from symbol to marker
- I made a component `VoronoiArea` containing `path` (we discussed making a mixin but I like this solution)
- I created a marker in the form of a star
- The form of the marker can be defined for each data point seperately. If less markers are specified than there are data points, the last specified marker is used for the remaining data points (eg if only one marker is specified, this marker is used for all data points). However, if multiple series share a data point, some points might get the wrong marker. The other properties of the marker (eg size) can be set per series. I'm not happy with the code structure in `DataSeries.jsx` because I'd like to be able to set all properties for each data point seperately. But I didn't figure out a good solution. It would be necessary to access the attributes of each data point when creating `VoronoiContainer`. Although some further work is necessary, it is working.

**Comment by MatthewHerbst** [comment]
Should these changes be applied to ScatterChart as well?

**Comment by nightlyop** [comment]
I think so but unfortunately I don't have time for that. I'm trying to add a possibility to display a grid now.

**Comment by MatthewHerbst** [comment]
Mmm. And I'm working on #237 . Maybe I'll have the time to make these changes to ScatterChart after I finish that. And I agree, the differing structures of `DataSeries` bothers me as well. We had some discussion about it in #219 but haven't talked about it in a while.

**Comment by nightlyop** [comment]
I saw that discussion after my PR. However I think it is quite a lot of work to figure out a good architecture and it might be better if someone with more experience (and time) could do that somewhen.

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,6 @@ build/
 
 # OSX
 .DS_Store
+dist/public/js/react-d3.min.js
+dist/public/js/react-d3.min.js.map
+test.js
```
Comment (yang-wei) [inline .gitignore]: what's `test.js` for. If you wish to ignore these file I think it's better to do it in another pr.

Comment (nightlyop) [inline .gitignore]: ugh, I was a little confused. I thought that's from the master. Don't know where that's from. I'll remove it.

**Comment by nightlyop** [comment]
The different kind markers can be used for scattercharts as well. I haven't touched the legend so far and I don't think I'll have time therefore in the next weeks. But if you merge the code I can write a help page.

**Comment by yang-wei** [comment]
:+1 will look at these later. need more time since it had so many files change

**Comment by nightlyop** [comment]
Sure, I'm not in a hurry.

**Comment by nightlyop** [comment]
Might have been better to make a new branch for this fix.. Sorry. Shall I make one?

**Comment by yang-wei** [comment]
yes ! If you could please squash the bugfix commit. thanks

**Comment by nightlyop** [comment]
Aaaw, sorry for the git mess. When I noticed that I forgot the `ScatterChart` I tried to amend the changes to the first commit but it didn't work as I expected. :-/

**Comment by nightlyop** [comment]
I wrote https://github.com/esbullington/react-d3/wiki/Marker about what I made in this branch. It's

**Comment by yang-wei** [comment]
@nightlyop Thanks for the hard work but I really think that our code base will become complicated after merging this. Give me some time, I might do a little bit refactoring before merging this into master.

File: src/common/marker/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Created by cfaes on 24.08.2015.
```
Comment (yang-wei) [inline src/common/marker/index.js]: Is your id cfaes ? We don't give credit in our code base. We do it in changelog or other place. Is it ok for me to remove this?

**Comment by yang-wei** [comment]
See issue #261

File: src/common/marker/index.js (hunk 2 above)
Comment (nightlyop) [inline src/common/marker/index.js]: Sure. That's added by webstorm automatically and I forgot to remove it.

**Comment by yang-wei** [comment]
Will close this due to inactive.
